Trace/src/Connection/Grpc.php

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Trace\Connection;
+
+use Google\ApiCore\Serializer;
+use Google\Cloud\Core\GrpcRequestWrapper;
+use Google\Cloud\Core\GrpcTrait;
+use Google\Cloud\Trace\TraceClient;
+use Google\Cloud\Trace\V2\TraceServiceClient;
+use Google\Cloud\Trace\V2\Span;
+
+/**
+ * Implementation of the
+ * [Google Trace gRPC API](https://cloud.google.com/trace/docs/).
+ */
+class Grpc implements ConnectionInterface
+{
+    use GrpcTrait;
+
+    /**
+     * @var Serializer
+     */
+    private $serializer;
+
+    /**
+     * @var TraceServiceClient
+     */
+    private $traceClient;
+
+    /**
+     * @param array $config
+     */
+    public function __construct(array $config = [])
+    {
+        $this->serializer = new Serializer();
+        $config['serializer'] = $this->serializer;
+        $this->setRequestWrapper(new GrpcRequestWrapper($config));
+        $gaxConfig = $this->getGaxConfig(
+            TraceClient::VERSION,
+            isset($config['authHttpHandler'])
+                ? $config['authHttpHandler']
+                : null
+        );
+
+        $this->traceClient = new TraceServiceClient($gaxConfig);
+    }
+
+    /**
+     * Sends new spans to new or existing traces. You cannot update existing
+     * spans.
+     *
+     * @param array $args {
+     *      Batch write params.
+     *
+     *      @type string $projectsId The ID of the Google Cloud Project
+     *      @type array $spans
+     * }
+     */
+    public function traceBatchWrite(array $args)
+    {
+        $spans = $this->pluck('spans', $args);
+        return $this->send([$this->traceClient, 'batchWriteSpans'], [
+            TraceServiceClient::projectName($this->pluck('projectsId', $args)),
+            array_map([$this, 'buildSpan'], $spans),
+            $args
+        ]);
+    }
+
+    /**
+     * @param array $args
+     */
+    public function traceSpanCreate(array $args)
+    {
+        return $this->send([$this->traceClient, 'createSpan'], [
+            TraceServiceClient::projectName($this->pluck('projectsId', $args)),
+            $this->pluck('spanId', $args),
+            $this->pluck('displayName', $args),
+            $this->formatTimestampForApi($this->pluck('startTime', $args)),
+            $this->formatTimestampForApi($this->pluck('endTime', $args)),
+            $args
+        ]);
+    }
+
+    private function buildSpan(array $span)
+    {
+        if (isset($span['startTime'])) {
+            $span['startTime'] = $this->formatTimestampForApi($span['startTime']);
+        }
+        if (isset($span['endTime'])) {
+            $span['endTime'] = $this->formatTimestampForApi($span['endTime']);
+        }
+        if (isset($span['timeEvents'])) {
+            foreach ($span['timeEvents']['timeEvent'] as &$timeEvent) {
+                if (isset($timeEvent['time'])) {
+                    $timeEvent['time'] = $this->formatTimestampForApi($timeEvent['time']);
```

Comment:
@michaelbausor This is the ugliness that is required to fix nested timestamps before using the serializer. The alternative is to do the conversion manually here, but it's extremely verbose to do every field for every subtype.

Reply:
Yep, it is pretty rough to have to convert those nested timestamps manually. Does adding transformations to the decoding step in Serializer fix this completely? If so we should do it I think.

Reply:
The only issue I can possibly see is that the custom serializer rules apply globally to that serializer regardless of which protobuf type it's encoding/decoding. It's possible that in some large nested message, that there are two fields named the same thing, but should be serialized differently (I don't have a specific example but perhaps something generic like `type`).

Another possibility: rather than having the veneer objects serialize themselves to an array which is then converted to protobuf types, we could instead have them serialize themselves directly to their protobuf counterparts (and vice versa). I *believe* the nested arrays was previously required for interop between the Rest/Grpc connection types, but if we can convert all Rest connections into the Gapic ones then the protobuf messages would be the common structure.

Reply:
Right - the arguments in ConnectionInterface are optimized for the Rest connection: https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/Trace/src/Connection/Rest.php

This seems like something we could do API by API rather than all at once. @dwsupplee @jdpedrie what are your thoughts on having a ConnectionInterface that took protobuf objects as arguments, if the Rest transport was provided via the GAPIC client?

Reply:
Created a [gax PR](https://github.com/googleapis/gax-php/pull/170). Re: the shared name variable name, you can also specify the transformer by protobuf type, so you wouldn't be restricted to doing it by name.

Reply:
@michaelbausor,

We’ve been thinking a bit about we want to handle the Connection layer going forward. At times we have felt it may be best to remove connection classes altogether in favor of the GAPICs, at other times we see some value in segregating the transport level code from the clients as it can help to simplify the the client logic.

In regards to modifying the various `ConnectionInterface`s to more strictly accept protobuf types constructed in the client, John and I discussed this offline a bit.

Some downsides to this approach:

- Could it be faster to avoid serializing/deserializing the protobuf objects in certain cases (especially if a user can’t install the protobuf extension)?
- What about cases like Datastore, where we don’t have a GAPIC yet? Technically we could create a connection implementation that manually triggers gRPC requests with protobuf.

It almost seems the question boils down to this: *Do we want to build the veneers in a way where they can only wrap the GAPICs or should they be left open to any implementation?* We’re not hard sold on either direction yet, it may warrant a bit more of a discussion.

Reply:
With a bit of initial testing of writing some log entries, I am seeing calls sent over REST that do not use the generated protobuf objects being about 20% faster on average than those that do (the protobuf extension was disabled). The test is a bit naive, and there could be other reasons to explain the difference in speed beyond the protobuf serialization/deserialization. I'll do a little more playing around to see what else I can find.

https://gist.github.com/dwsupplee/3ec399fb43b30218ec08cb7846a72104

Reply:
@michaelbausor With the accepted PR (next version of gax), the serialization can be handled nicely with:

```php
$this->serializer = new Serializer(
            [],
            [
                'google.protobuf.Timestamp' => function ($v) {
                    return $this->formatTimestampFromApi($v);
                }
            ],
            [],
            [
                'google.protobuf.Timestamp' => function ($v) {
                    return $this->formatTimestampForApi($v);
                }
            ]
        );
$this->serializer->decodeMessage(new Span(), $span);
```